src/HieDb/Utils.hs

```diff
@@ -38,25 +38,39 @@ import Data.IORef
 import HieDb.Types
 import HieDb.Compat
 import Database.SQLite.Simple
+import Control.Monad.State.Strict (StateT, get, put)
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Control.Monad (when)
 
 #if __GLASGOW_HASKELL__ >= 903
 import Control.Concurrent.MVar (readMVar)
 #endif
 
-addTypeRef :: HieDb -> FilePath -> A.Array TypeIndex HieTypeFlat -> A.Array TypeIndex (Maybe Int64) -> RealSrcSpan -> TypeIndex -> IO ()
+-- Each AST Node can have a lot of repetitive type information.
+-- We use this to make sure that each occurrence of a type
+-- (Identified by Int64 id of that type within typenames table)
+-- is inserted exactly once per each RealSrcSpan in which it occurs and per
+-- depth of that type within the tree structure representing the type (Int)
+type TypeIndexing a = StateT (Set (RealSrcSpan, Int64, Int)) IO a
+
+addTypeRef :: HieDb -> FilePath -> A.Array TypeIndex HieTypeFlat -> A.Array TypeIndex (Maybe Int64) -> RealSrcSpan -> TypeIndex -> TypeIndexing ()
 addTypeRef (getConn -> conn) hf arr ixs sp = go 0
   where
     sl = srcSpanStartLine sp
     sc = srcSpanStartCol sp
     el = srcSpanEndLine sp
     ec = srcSpanEndCol sp
-    go :: TypeIndex -> Int -> IO ()
+    go :: Int -> TypeIndex -> TypeIndexing ()
```

Comment:
Correct me if I'm wrong, but I think that Int (depth) and TypeIndex were flipped.

Reply:
You are correct.